src/client_side_request.cc

```diff
@@ -2022,6 +2022,25 @@ ClientHttpRequest::handleAdaptedHeader(Http::Message *msg)
     } else if (HttpReply *new_rep = dynamic_cast<HttpReply*>(msg)) {
         debugs(85,3,HERE << "REQMOD reply is HTTP reply");
 
+        if (request->method == Http::METHOD_CONNECT) {
+            // CONNECT requests that REQMOD produces a reply for need special
+            // handling.
+            if (Http::Is2xx(new_rep->sline.status())) {
+                // REQMOD replying with a 2xx is a bug, since Squid won't have
+                // connected to the origin server.
+                debugs(85, DBG_IMPORTANT, HERE << "REQMOD produced a 2xx HTTP response for a CONNECT request");
+                static const auto d = MakeNamedErrorDetail("CLT_REQMOD_2xx_HTTP_RESPONSE");
+                handleAdaptationFailure(d, false);
+                return;
+            } else {
+                // Returning an error to the client, so turn readMore back on
+                // so we can keep the connection alive for the next request.
+                debugs(85,3,HERE << "REQMOD reply rejected a CONNECT");
+                ConnStateData * conn = getConn();
+                if (conn) conn->flags.readMore = 1;
```

Comment:
Yes, this is so that we don't get abandoned.  I think you're probably right that we should call mayUseConnection(false) here.  Oddly, I'm using (almost) this code in the wild on Squid 5 and it is working.

Reply:
Yeah, Stream::mayUseConnection_ documentation clearly implies that we should make mayUseConnection() false after we decide not to open a tunnel:

```C++
    bool mayUseConnection_; /* This request may use the connection. Don't read anymore requests for now */
```

I am not sure that documentation is 100% accurate, but I see no specific reasons not to trust it.

To reduce chances of this code getting out of sync again, I suggest adding a pair of functions or methods: reserveForTunnel() and cancelTunnelReservation() or perhaps unreserveForTunnel(). The existing "Let tunneling code be fully responsible for CONNECT requests" code will call the former, while your new code (and perhaps other code that decides not to tunnel?) will call the latter. What do you think?


>  Oddly, I'm using (almost) this code in the wild on Squid 5 and it is working.

Well, your "wild" may not be wild enough (or you just do not see some of the subtle failures). These flags become especially important when clients pipeline requests or do other relatively unusual/rare things. In the majority of use cases that I see in the code, mayUseConnection_ is checked before this PR code runs. ConnStateData::terminateAll() comes later, but it only runs in special cases.

Reply:
> To reduce chances of this code getting out of sync again, I suggest adding a pair of functions or methods: reserveForTunnel() and cancelTunnelReservation() or perhaps unreserveForTunnel(). The existing "Let tunneling code be fully responsible for CONNECT requests" code will call the former, while your new code (and perhaps other code that decides not to tunnel?) will call the latter. What do you think?

Sounds like a good plan.

> Well, your "wild" may not be wild enough (or you just do not see some of the subtle failures). These flags become especially important when clients pipeline requests or do other relatively unusual/rare things.

That's entirely possible.  In my case, my ICAP REQMOD service is responsible for deciding which requests to authenticate, and it returns a suitable "407" HTTP response(*).  Clients will regularly reuse the connection to send the authenticated request, but you're right that we're probably not seeing pipelined requests.

(* previously we were using an external ACL to talk to the ICAP server and control authentication, but eventually decided that we could dispense with it by just having the ICAP server produce the 407 response itself.  We've been using the new method for a few months now and it seems to work well.)